### PR TITLE
feat: added handlebar variable checker and defaults

### PR DESCRIPTION
**Pull request by venarius**
- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
This adds an component to the template editors where users can see their current handlebar variables (basic) in use, can specify defaults for them and if they are required. The api then enforces these rules.

  Screenshot 2022-08-29 053712

### CODE REVIEW

**Comment by venarius** [comment]
fyi, CI tests are currently failing because I did not run a ```pnpm install``` yet as the main branch ```pnpm-lock.yaml``` seems to have been downgraded to lockfile version ```5.3``` and I don't know if that's intentional.

File: libs/shared/src/entities/message-template/message-template.interface.ts [hunk 1]
```diff
@@ -9,13 +9,21 @@ export interface IEmailBlock {
   };
 }
 
+export class ITemplateVariable {
+  type: 'String' | 'Array' | 'Boolean';
```
Comment (scopsy) [inline libs/shared/src/entities/message-template/message-template.interface.ts:13]: Maybe we can create a type out of it? 

```
type TemplateVariableType = 'String' | 'Array' | 'Boolean';
```

Or even an enum. And we can have it in the @novu/shared to be reused between client and backend

File: apps/web/src/components/templates/email-editor/EmailContentCard.tsx [hunk 2]
```diff
@@ -79,6 +80,7 @@ export function EmailContentCard({
   return (
     <>
       {!isIntegrationActive ? <LackIntegrationError channelType="E-Mail" /> : null}
+      <VariableManager template={`steps.${index}.template`} />
```
Comment (scopsy) [inline apps/web/src/components/templates/email-editor/EmailContentCard.tsx:83]: Maybe in the meanwhile let's put it under each content area and not above? Will consult with the designer about this one, but for now I think it can be below. Wdyt?

File: apps/web/src/components/templates/chat-editor/TemplateChatEditor.tsx [hunk 3]
```diff
@@ -24,6 +25,7 @@ export function TemplateChatEditor({
   return (
     <>
       {!isIntegrationActive ? <LackIntegrationError channelType={ChannelTypeEnum.CHAT} /> : null}
+      <VariableManager template={`steps.${index}.template`} />
```
Comment (scopsy) [inline apps/web/src/components/templates/chat-editor/TemplateChatEditor.tsx:28]: A small question, is the variables shared across all the different templates?

File: apps/web/src/components/templates/VariableManager.tsx [hunk 4]
```diff
@@ -0,0 +1,209 @@
+import { IEmailBlock } from '@novu/shared';
```
Comment (scopsy) [inline apps/web/src/components/templates/VariableManager.tsx:1]: Do you think we can add `.cy.tsx` component tests here to test the edgecases here in isolation?

File: apps/web/src/components/templates/VariableManager.tsx [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+  type: MustacheVariableType;
+  name: string;
+  defaultValue?: string | boolean;
+  required?: boolean;
+}
+
+export const VariableComponent = ({ index, template }: VariableComponentProps) => {
+  const { control, watch } = useFormContext();
+
+  const variableName = watch(`${template}.variables.${index}.name`);
```
Comment (scopsy) [inline apps/web/src/components/templates/VariableManager.tsx:32]: When looking at the email template, we also have custom code support. For now looks like it's not watching the variables from there and only on the visual editor

File: apps/api/src/app/events/usecases/verify-payload/verify-payload.usecase.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  execute(command: VerifyPayloadCommand): Record<string, unknown> {
+    const invalidKeys = [];
+    const defaultPayload = {};
+
+    for (const step of command.template.steps) {
+      invalidKeys.push(...this.checkRequired(step.template.variables || [], command.payload));
+    }
+
+    if (invalidKeys.length)
+      throw new ApiException(`payload is missing required key(s) and type(s): ${invalidKeys.join(', ')}`);
```
Comment (scopsy) [inline apps/api/src/app/events/usecases/verify-payload/verify-payload.usecase.ts:15]: Great Error!

File: apps/api/src/app/events/usecases/verify-payload/verify-payload.usecase.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    return invalidKeys;
+  }
+
+  private fillDefaults(variables: ITemplateVariable[], payload: Record<string, unknown>) {
+    for (const variable of variables.filter((vari) => vari.defaultValue !== undefined && vari.defaultValue !== null)) {
+      this.setNestedKey(payload, variable.name.split('.'), variable.defaultValue);
+    }
+  }
```
Comment (scopsy) [inline apps/api/src/app/events/usecases/verify-payload/verify-payload.usecase.ts:66]: A small suggestion, is it possible to have those function to return values instead of modifying a reference of an object? In that case it might be easier to comprehend what's going on.

File: apps/api/src/app/events/usecases/verify-payload/verify-payload.usecase.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
+import { ITemplateVariable } from '@novu/dal';
+import { ApiException } from '../../../shared/exceptions/api.exception';
+import { VerifyPayloadCommand } from './verify-payload.command';
+
+export class VerifyPayload {
+  execute(command: VerifyPayloadCommand): Record<string, unknown> {
```
Comment (scopsy) [inline apps/api/src/app/events/usecases/verify-payload/verify-payload.usecase.ts:6]: Could we create a unit tests for this function to make sure all edge cases are covered here?

File: apps/api/src/app/events/usecases/verify-payload/verify-payload.usecase.ts [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    const invalidKeys = [];
+    const defaultPayload = {};
+
+    for (const step of command.template.steps) {
+      invalidKeys.push(...this.checkRequired(step.template.variables || [], command.payload));
```
Comment (scopsy) [inline apps/api/src/app/events/usecases/verify-payload/verify-payload.usecase.ts:11]: This is amazing, but there are a few variables we are passing to the template:
- payload
- subscriber (the current subscriber we are sending it to)
- step (this comes from the digest)

Those variables are injected to the template compilation but are not in the command.payload. So we need to create a combined variables object containing all of the above.

File: apps/api/src/app/events/usecases/trigger-event/trigger-event.usecase.ts [hunk 10]
```diff
@@ -49,6 +52,15 @@ export class TriggerEvent {
...
     }
 
+    const defaultPayload = this.verifyPayload.execute(
+      VerifyPayloadCommand.create({
+        payload: command.payload,
+        template,
+      })
+    );
+
+    command.payload = { ...command.payload, ...defaultPayload };
```
Comment (scopsy) [inline apps/api/src/app/events/usecases/trigger-event/trigger-event.usecase.ts:62]: In the case of nested object in one or the other, won't they cancel each other?

File: apps/api/src/app/events/e2e/trigger-event.e2e.ts [hunk 11]
```diff
@@ -475,6 +475,158 @@ describe('Trigger event - /v1/events/trigger (POST)', function () {
...
+            lastName: '',
+            array: [],
+            bool: true,
+          },
+        },
+      })
+      .expect(201);
+  });
+
+  it('should fill trigger payload with default variables', async function () {
```
Comment (scopsy) [inline apps/api/src/app/events/e2e/trigger-event.e2e.ts:548]: Great test!

**Review by scopsy — COMMENTED** [review]
@venarius this is just amazing, This is an incredible effort to make the experience so much better now! I've added a few comments (Sorry about that) but mainly for a few edge cases and etc...

Let me know what you think

File: apps/web/src/components/templates/chat-editor/TemplateChatEditor.tsx (hunk 3 above)
Comment (venarius) [reply apps/web/src/components/templates/chat-editor/TemplateChatEditor.tsx:28]: No, it's bound to each step.

File: apps/api/src/app/events/usecases/verify-payload/verify-payload.usecase.ts (hunk 9 above)
Comment (venarius) [reply apps/api/src/app/events/usecases/verify-payload/verify-payload.usecase.ts:11]: The other two variables are currently filled by the app itself and not the user. I would keep the current functionality only to verify the custom event payload and not the fields generated by the app. Wdyt?

File: apps/api/src/app/events/usecases/trigger-event/trigger-event.usecase.ts (hunk 10 above)
Comment (venarius) [reply apps/api/src/app/events/usecases/trigger-event/trigger-event.usecase.ts:62]: True, nice finding! I thought the ES6 spread nowadays does more than just a shallow merge haha

File: apps/api/src/app/events/usecases/verify-payload/verify-payload.usecase.ts (hunk 9 above)
Comment (venarius) [reply apps/api/src/app/events/usecases/verify-payload/verify-payload.usecase.ts:11]: I would rather mark these variables as system reserved if the user uses them in the template. Because currently the user can just override them and this would case some unexpected issues. Wdyt?

Screenshot 2022-08-31 230854

File: apps/web/src/components/templates/chat-editor/TemplateChatEditor.tsx (hunk 3 above)
Comment (scopsy) [reply apps/web/src/components/templates/chat-editor/TemplateChatEditor.tsx]: Awesome this makes a lot of sense. Maybe later we could also add a global view of the combined variables from all inputs on the workflow editor. But for now it's totally perfect.

File: apps/api/src/app/events/usecases/verify-payload/verify-payload.usecase.ts (hunk 9 above)
Comment (scopsy) [reply apps/api/src/app/events/usecases/verify-payload/verify-payload.usecase.ts:12]: yes this makes perfect sense. Maybe we can do a less in your face message there, because it's not an error but an info. Maybe doing it with a bright grey color instead of red?

File: apps/web/src/components/templates/TemplateSMSEditor.tsx [hunk 12]
```diff
@@ -40,6 +41,7 @@ export function TemplateSMSEditor({
           />
         )}
       />
+      <VariableManager index={index} contents={['content']} />
```
Comment (scopsy) [inline apps/web/src/components/templates/TemplateSMSEditor.tsx:44]: Love the API!

File: apps/api/src/app/events/usecases/verify-payload/verify-payload.usecase.ts (hunk 9 above)
Comment (venarius) [reply apps/api/src/app/events/usecases/verify-payload/verify-payload.usecase.ts:12]: Sure, I changed it :)

**Review by scopsy — APPROVED** [review]
Perfecto!

**Comment by scopsy** [comment]
This test 😅🤦‍♂️

**Comment by scopsy** [comment]
@djabarovgeorge @davidsoderberg could you please take a final look before I merge this to the release aswell? 🙏

**Comment by davidsoderberg** [comment]
Looks awesome 😍

**Comment by scopsy** [comment]
Merging this 🚀
